### PR TITLE
Fpd 2.0 Update

### PR DESCRIPTION
**Pull request by mmoschovas**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [X ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

DETAIL:
Prebid's First Party Data (FPD) convention revision with backwards compatibility. 

MODULES UPDATED:
gptPreAuction, prebidServer, admixer, adrelevantis, amx, avocet, criteo, grid, haloRtd, imar, jwPlayerRtd, luponmedia, rubicon, smaato, triplelift

BACKWARD COMPATIBILITY GLOBAL/BIDDER:
fpd.context.ATTR --> ortb2.site.ATTR
fpd.context.data.ATTR --> ortb2.site.ext.data
fpd.user.ATTR --> ortb2.user.ATTR
fpd.user.data.ATTR --> ortb2.user.ext.data. (for fpd only... ortb2 can specify user.data)

BACKWARD COMPATIBILITY ADUNIT:
AdUnit.fpd - > AdUnit.ortb2Imp
AdUnit.fpd.context.ATTR -> AdUnit.ortb2Imp.ext.data.ATTR

ISSUE:
https://github.com/prebid/Prebid.js/issues/5795

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by bretg** [comment]
@anthonylauzon - halo is the only module not in this PR... what you're doing wasn't clear to us. Happy to sit with someone on your team to describe generally what needs to be done so you guys can make the appropriate changes. 

I'll be reaching out via email to as many of the affected parties as I can find addresses for so they can confirm and do any additional testing. In the meantime, here's a list of github handles that have recently touched affected modules

admixer @AdmixerTech
amx @nickjacob
avocet @scritchley
inmar @zandree-owneriq
criteo @leonardlabat , @allanjun
grid @TheMediaGrid
luponmedia @adxpremium
smaato @sbrosinski
triplelift @kzhouTL , @colbertk , @davidwoodsandersen

**Comment by anthonylauzon** [comment]
@bretg i will pull this branch and do what i can to update our module to meet the new spec. we'll set up a call if there are any issues requiring discussion.  i'll request you as a reviewer when that's complete.

**Comment by bretg** [comment]
I've sent direct email to: admixer, adrelevantis, amx, avocet, criteo, grid, inmar, jwplayer, luponmedia, smaato, and triplelift.

**Comment by leonardlabat** [comment]
Hello @bretg 
It looks ok for the criteo adapter 👍 
Thanks for the change

**Comment by AdmixerTech** [comment]
Hi @bretg 
looks good for admixer adapter
thanks

File: src/prebid.js [hunk 1]
```diff
@@ -840,7 +840,7 @@ $$PREBID_GLOBAL$$.markWinningBidAsUsed = function (markBidRequest) {
  */
 $$PREBID_GLOBAL$$.getConfig = config.getConfig;
 
-/**
+/** ƒ
```
Comment (karimMourra) [inline src/prebid.js:843]: nit: typo

Comment (mmoschovas) [reply src/prebid.js]: Remove this. Thanks for catching

**Review by karimMourra — APPROVED** [review]
Approving changes made to the JW Player RTD module! Thank you so much for the update!

**Review by TheMediaGrid — APPROVED** [review]

**Comment by Enigo** [comment]
Approving changes made to the **Smaato** module.
thanks, guys, we will remove the legacy code soon.

**Comment by bretg** [comment]
@musikele - Jason won't be able to get to this until next week. The other reviewers only verified their individual adapter changes. Could you take a look at the PBS-core, gptPreAuction, pbsBidAdapter, and Rubicon changes here?

**Comment by anthonylauzon** [comment]
As noted in https://github.com/prebid/Prebid.js/issues/5795, I am not seeing how it is possible to add dynamic segment information for users within the ortb object defined by setConfig.  I can add this data at an ad unit level, but this would result in duplication and the bid adapters might not be looking for it.

**Comment by bretg** [comment]
@anthonylauzon - I agree that the setConfig interface is awkward for RTD modules.

In the short term, the suggestion is that the RTD module calls getConfig({ortb2}), updates the output, and then calls setConfig({ortb2}).

I don't know how to handle bidder-sensitivity here. 
1) I suspect we're going to need to extend the 'eidPermissions' feature to 'rtdPermissions'
2) I don't think setBidderConfig() will handle "merging" multiple configs properly, so an initial hack may be needed.

**Comment by bretg** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/2722

**Review by jsnellbaker — CHANGES_REQUESTED** [review]
Hi @mmoschovas 

Thanks for putting everything together.  I think the core changes are mostly in good shape.  There was one potential issue I observed during testing; details below:

Supplied value:
```
      pbjs.setConfig({
        ortb2: {
          site: {
            name: 'example-name',
            domain: 'page.example.com',
            cat: ['IAB2'],
            sectioncat: ["IAB2-2"],
            pagecat: ["IAB2-2"],
            page: "https://page.example.com/here.html",
            ref: "https://ref.example.com",
            keywords: 'power tools, drills',
            search: 'drill',
            content: { userrating: 4 },
            ext: {
              data: {
                pageType: 'article',
                category: 'repair'
              }
            }
          },
          user: {
            yob: 1945,
            gender: 'm',
            keywords: 'a,b',
            data: [{
              name: 'dataprovider.com',
              ext: { taxonomyname: 'IAB' },
              segment: [{
                id: '1'
              }]
            }],
            ext: {
              data: {
                registered: true,
                interests: ['cars']
              }
            }
          }
        }
      });
```
converted value from config.getLegacyFpd():
```
{
  "context": {
    "name": "example-name",
    "domain": "page.example.com",
    "cat": [
      "IAB2"
    ],
    "sectioncat": [
      "IAB2-2"
    ],
    "pagecat": [
      "IAB2-2"
    ],
    "page": "https://page.example.com/here.html",
    "ref": "https://ref.example.com",
    "keywords": "power tools, drills",
    "search": "drill",
    "content": {
      "userrating": 4
    },
    "data": {
      "pageType": "article",
      "category": "repair"
    }
  },
  "user": {
    "yob": 1945,
    "gender": "m",
    "keywords": "a,b",
    "data": [
      {
        "name": "dataprovider.com",
        "ext": {
          "taxonomyname": "IAB"
        },
        "segment": [
          {
            "id": "1"
          }
        ]
      }
    ]
  }
}
```

If you look at the results in the legacy converted user object, it populated the 'old' `data` field from the `ortb2.user.data` array field rather than `ortb2.user.ext.data` (which was the old `fpd.user.data` field).

The conversion works fine if your setConfig uses the fpd setup and the code converts it forward and then backward.  This is likely b/c of the shared name `data` wasn't in the fpd construct.  

Can you please take a look into this?  I think based on what we discussed the other day, we need to fix this behavior.

Please let me know if you have questions about the above. Thanks!

**Comment by bretg** [comment]
Good catch @jsnellbaker - I didn't test config.getLegacyFpd() - hopefully pretty easy fix @mmoschovas ?

**Review by ncolletti — COMMENTED** [review]
This looks good so far @mmoschovas, nice work. Only have a few small comments from a first pass, will go over more with tests tomorrow.

**Comment by mmoschovas** [comment]
Thanks for reviewing @jsnellbaker and @ncolletti. Just pushed changes that should cover the above requests

**Review by ncolletti — APPROVED** [review]
LGTM

**Review by jsnellbaker — APPROVED** [review]
@mmoschovas  I think this looks good now; thanks for making the fix.

@bretg Are there any more reviews we're waiting on from the bidders who are being updated?  Or should we be free to merge?

**Comment by ncolletti** [comment]
Noticed the PR_REVIEW markdown references fpd.context and fpd.user, may want to update that here.

**Comment by bretg** [comment]
Great news. Ok, I've upgraded this to a "feature", and plan to modify the release notes to point to the First Party Data page. 

Merging.

**Pull request merged**